bin/reflective_loader/src/loader.rs

```diff
@@ -0,0 +1,844 @@
+fn virtual_alloc(fn_ptr: FnVirtualAlloc, err_ptr: FnGetLastError, lp_address: *const c_void, dw_size: usize, fl_allocation_type: u32, fl_protect: u32) -> *mut c_void {
+    let buffer_handle: *mut c_void = unsafe{ fn_ptr(lp_address, dw_size, fl_allocation_type, fl_protect) };
+    if buffer_handle == null_mut() {
+        let error_code = unsafe { err_ptr() };
+        if error_code != 0 {
+            panic!("Failed to allocate memory. Last error returned: {}", error_code);
+        }
+    }
+    buffer_handle
+}
+
+// pub unsafe fn GetProcAddress(hmodule: P0, lpprocname: P1) -> FARPROC
+fn get_proc_address(fn_ptr: FnGetProcAddress, err_ptr: FnGetLastError, h_module: isize, lp_proc_name: *const u8) -> isize {
+    let proc_handle = match unsafe { fn_ptr(h_module, lp_proc_name) } {
+        Some(local_proc_handle) => local_proc_handle,
+        None => {
+            let error_code = unsafe { err_ptr() };
+            panic!("Failed to find function {:?} in module. Last error returned: {}", lp_proc_name, error_code);
+        }
+    };
+    proc_handle as isize
+}
+
+
+// pub unsafe fn LoadLibraryA(lplibfilename: P0) ->  Result<HMODULE>
+fn load_library_a(fn_ptr: FnLoadLibraryA, err_ptr: FnGetLastError, lplibfilename: PCSTR) -> HINSTANCE {
+    let library_handle = unsafe { fn_ptr(lplibfilename) };
+    if library_handle == 0 {
+        let error_code = unsafe { err_ptr() };
+        if error_code != 0 {
+            panic!("Failed to load library. Last error returned: {}", error_code);
+        }
+    }
+    library_handle
+}
+
+#[derive(Debug, Copy, Clone)]
+struct BaseRelocationEntry {
+    offset: u16,
+    reloc_type: u16,
+}
+
+impl BaseRelocationEntry {
+    fn new(c_bytes: u16) -> Self {
+        let reloc_type_bit_mask: u16 = 0b1111_0000_0000_0000;
+        let reloc_type = (c_bytes & reloc_type_bit_mask) >> 12;
+        
+        let offset_bit_mask: u16 = 0b0000_1111_1111_1111;
+        let offset = c_bytes & offset_bit_mask;
+        
+        Self {
+            offset,
+            reloc_type,
+        }
+    }
+
+    fn c_size() -> usize {
+        core::mem::size_of::<u16>()
+    }
+}
+
+#[cfg(target_arch = "x86_64")]
+struct PeFileHeaders64 {
+    dos_headers: IMAGE_DOS_HEADER,
+    nt_headers: IMAGE_NT_HEADERS64,
+    section_headers: [IMAGE_SECTION_HEADER; MAX_PE_SECTIONS],
+}
+
+// Pares the PE file from a series of bytes
+#[cfg(target_arch = "x86_64")]
+impl PeFileHeaders64 {
+    fn new(dll_bytes_ptr: *mut c_void) -> Self{
+        // DOS Headers
+        let dos_headers_base_ptr = dll_bytes_ptr as usize;
+        let dos_headers = unsafe { *((dos_headers_base_ptr) as *mut IMAGE_DOS_HEADER) };
+        if dos_headers.e_magic != PE_MAGIC {
+            panic!("PE Magic header mismatch. Expected 0x5A4D == MZ == 21117. File does not appear to be a PE executable.");
```

Comment:
I really hate panic. Make a Result type and pass the error up.

Reply:
In the loader i'm using panic for two reasons:
- Bubbling errors requires `memcpy` which would import the function from the C runtime. I've re-implemented `memcpy` in the loader since one specific operation (copying bytes between buffers) keeps being compiled down to `memcpy`. If possible though i'd like to avoid relying on the `memcpy` function I wrote.
- The agent doesn't have a way to receive errors from the loader since it's copied into the remote process and executed. Any errors in the loader should abort.

Gonna leave this how it is for now since I can't benefit from the Rust errors and it could introduce reliability issues with my non battle tested memcpy implementation.

Reply:
https://github.com/KCarretto/realm/pull/205#discussion_r1243099512
- I'll look into this if we can get a real version of memcpy erroring would be worth IMO.